File: Leetcode 230 - kth-smallest-element-in-a-bst/230.kth-smallest-element-in-a-bst.cpp
/*
 * @lc app=leetcode id=230 lang=cpp
 *
 * [230] Kth Smallest Element in a BST
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
  int kthSmallest(TreeNode *root, int k) {
    vector<int> tree;
    queue<TreeNode *> q;
    q.push(root);

    while (!q.empty()) {
      int len = q.size();

      while (len--) {
        TreeNode *cur = q.front();
        q.pop();

        if (cur) {
          q.push(cur->left);
          q.push(cur->right);
          tree.push_back(cur->val);
        }
      }
    }

    sort(tree.begin(), tree.end());

    return tree[k - 1];
  }
};
// @lc code=end

File: Leetcode 230 - kth-smallest-element-in-a-bst/230.kth-smallest-element-in-a-bst.java
/*
 * @lc app=leetcode id=230 lang=java
 *
 * [230] Kth Smallest Element in a BST
 */

// @lc code=start
import java.util.ArrayList;
import java.util.List;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

  // According to the definition of a BST, using dfs, the smallest element of the tree will be first put inside the list
  // So we can just return the kth element in the inorder traversal
  public int kthSmallest(TreeNode root, int k) {
    List<Integer> inorder = new ArrayList<>();

    dfs(root, inorder);

    return inorder.get(k - 1);
  }

  private void dfs(TreeNode root, List<Integer> inorder) {
    if (root == null) return;

    dfs(root.left, inorder);

    inorder.add(root.val);

    dfs(root.right, inorder);
  }
}
// @lc code=end
